File: src/main/java/com/plugin/frege/resolve/FregeNativeFunctionNameReference.java
package com.plugin.frege.resolve;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.plugin.frege.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static com.plugin.frege.psi.impl.FregePsiClassUtilImpl.getClassesByQualifiedName;
import static com.plugin.frege.psi.impl.FregePsiClassUtilImpl.getMethodsAndFieldsByName;

public class FregeNativeFunctionNameReference extends FregeReferenceBase {
    public FregeNativeFunctionNameReference(@NotNull PsiElement element) {
        super(element, new TextRange(0, element.getTextLength()));
    }

    // TODO take into account: signatures
    @Override
    protected List<PsiElement> resolveInner(boolean incompleteCode) {
        FregeNativeFun nativeFunction = PsiTreeUtil.getParentOfType(element, FregeNativeFun.class);
        if (nativeFunction == null) {
            return List.of();
        }

        List<String> nativeNames;
        String methodName;
        PsiElement nativeNameFromJavaItem = getNativeNameFromJavaItem(nativeFunction.getJavaItem());
        if (nativeNameFromJavaItem != null) {
            String nativeNameString = nativeNameFromJavaItem.getText();
            if (nativeNameString.contains(".")) {
                nativeNames = List.of(nativeNameString.substring(0, nativeNameString.lastIndexOf('.')));
                methodName = nativeNameString.substring(nativeNameString.lastIndexOf('.') + 1);
            } else {
                nativeNames = List.of();
                methodName = nativeNameString;
            }
        } else {
            List<FregeSigma> sigmas = nativeFunction.getSigmaList();
            if (sigmas.isEmpty()) {
                return List.of();
            }

            FregeSigma sigma = sigmas.get(0);
            FregeDataNameUsage dataNameUsage = PsiTreeUtil.findChildOfType(sigma, FregeDataNameUsage.class);
            nativeNames = getNativeNamesFromDataNameUsage(dataNameUsage, incompleteCode);
            methodName = element.getText();
        }

        Project project = element.getProject();
        return nativeNames.stream()
                .flatMap(name -> getClassesByQualifiedName(project, name).stream())
                .filter(Objects::nonNull)
                .flatMap(psiClass -> getMethodsAndFieldsByName(psiClass, methodName).stream())
                .collect(Collectors.toList());
    }

    private @NotNull List<String> getNativeNamesFromDataNameUsage(@Nullable FregeDataNameUsage dataNameUsage,
                                                                      boolean incompleteCode) {
        if (dataNameUsage == null) {
            return List.of();
        }
        return new FregeDataNameUsageReference(dataNameUsage).resolveInner(incompleteCode).stream()
                .map(this::getNativeNameFromData)
                .filter(Objects::nonNull)
                .map(PsiElement::getText)
                .collect(Collectors.toList());
    }

    private @Nullable PsiElement getNativeNameFromJavaItem(@Nullable FregeJavaItem javaItem) {
        if (javaItem == null) {
            return null;
        }
        return javaItem.getNativeName();
    }

    private @Nullable FregeNativeName getNativeNameFromData(@NotNull PsiElement dataName) {
        if (!(dataName instanceof FregeDataNameNative)) {
            return null;
        }
        FregeDataDclNative dataNative = PsiTreeUtil.getParentOfType(dataName, FregeDataDclNative.class);
        if (dataNative == null) {
            return null;
        }
        return PsiTreeUtil.getChildOfType(dataNative, FregeNativeName.class);
    }
}

File: src/main/kotlin/com/plugin/frege/resolve/FregeNativeFunctionNameReference.kt
package com.plugin.frege.resolve

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.util.parentOfTypes
import com.plugin.frege.psi.*
import com.plugin.frege.psi.impl.FregePsiUtilImpl
import com.plugin.frege.resolve.FregeResolveUtil.findClassesByQualifiedName
import com.plugin.frege.resolve.FregeResolveUtil.findMethodsAndFieldsByName

class FregeNativeFunctionNameReference(element: PsiElement) : FregeReferenceBase(element, TextRange(0, element.textLength)) {

    // TODO take into account: signatures
    override fun resolveInner(incompleteCode: Boolean): List<PsiElement> { // TODO support incomplete code
        val nativeFunction = psiElement.parentOfTypes(FregeNativeFunction::class) ?: return emptyList()

        val nativeNames: List<String>
        val methodName: String
        val nativeNameFromJavaItem = getNativeNameFromJavaItem(nativeFunction.javaItem)
        if (nativeNameFromJavaItem != null) {
            val nativeNameString = nativeNameFromJavaItem.text

            methodName = FregePsiUtilImpl.nameFromQualifiedName(nativeNameString)
            nativeNames = if (nativeNameString.contains(".")) {
                listOf(FregePsiUtilImpl.qualifierFromQualifiedName(nativeNameString))
            } else {
                emptyList()
            }
        } else {
            val sigmas = nativeFunction.sigmaList
            if (sigmas.isEmpty()) {
                return emptyList()
            }

            val sigma = sigmas[0]
            val dataNameUsage = PsiTreeUtil.findChildOfType(sigma, FregeConidUsage::class.java)
            nativeNames = getNativeNamesFromDataNameUsage(dataNameUsage, incompleteCode)
            methodName = psiElement.text
        }

        val project = psiElement.project
        return nativeNames.asSequence()
            .flatMap { findClassesByQualifiedName(project, it) }
            .flatMap { findMethodsAndFieldsByName(it, methodName) }
            .toList()
    }

    private fun getNativeNamesFromDataNameUsage(
        dataNameUsage: FregeConidUsage?,
        incompleteCode: Boolean
    ): List<String> {
        return if (dataNameUsage != null) {
            FregeConidUsageReference(dataNameUsage).resolveInner(incompleteCode)
                .mapNotNull { getNativeNameFromData(it)?.text }
        } else {
            emptyList()
        }
    }

    private fun getNativeNameFromJavaItem(javaItem: FregeJavaItem?): PsiElement? {
        return javaItem?.nativeName
    }

    private fun getNativeNameFromData(dataNative: PsiElement): FregeNativeName? {
        return if (dataNative is FregeNativeDataDecl) {
            PsiTreeUtil.getChildOfType(dataNative, FregeNativeName::class.java)
        } else {
            null
        }
    }
}
